### PR TITLE
New tool to prepare ginan to run using a rinex file as input

### PR DESCRIPTION
**Pull request by aclel**
This PR proposes to introduce a new tool to help users when running Ginan with their own RINEX files for the first time.

* It parses the RINEX file to extract information from the header and observations
* It downloads the required IGS products to run Ginan for a ppp static solution
* It generates a yaml config for Ginan to be able to get a ppp static solution

These three functions are unified in a `prep` command. This command currently has a narrow use case - to allow Ginan to run RINEX files that would normally be submitted to state government for inclusion in the national adjustment. The intention is to enhance the tool's capabilities over time to support more use cases.

Please suggest a better name than `gn` - I wanted something that wasn't `ginan` and it's just what came to mind. Maybe there is an aboriginal word in the same realm as `ginan` that would have some significance?

### CODE REVIEW

File: scripts/auto_generate_yaml.py
Comment (ronaldmaj) [inline scripts/auto_generate_yaml.py:1]: Just checking, you didn't end up using this file in your implementation right?

File: scripts/gn/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+### Ginan Python CLI Tools
+
+This is a collection of tools to help you prepare your own data to run with Ginan.
+
+## Getting started
+
+Ensure you have created a virtual environment in the `scripts` folder and installed the `requirements.txt`.
```
Comment (ronaldmaj) [inline scripts/gn/README.md:7]: We could include instructions on how to do this via venv for example (or point people to the relevant docs that explain how to set up a virtual env)

File: scripts/gn/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+Also ensure you have run `python3 scripts/s3_filehandler.py -d -p` to populate the `inputData` folder with the constants that `pea` depends on.
```
Comment (ronaldmaj) [inline scripts/gn/README.md:9]: I would like to replace this with the `auto_download_PPP` script. I think I've got most (all?) of the relevant inputs now covered, especially following the new ver 3 ppp_example template. 
I think for now this is ok, but we'll need to do some testing seeing how that script can be incorporated here

File: scripts/gn/download_rinex_deps.py
Comment (ronaldmaj) [inline scripts/gn/download_rinex_deps.py:1]: Some of this is now covered in the `auto_download_PPP` script, so we'll need to look at merging what is done here with the use of that script (or at least some of the functions there)
Again, what you've got here is good, and I think it's fine to go in, but we'll need to have a closer look at that in the next step

File: scripts/gn/parse_rinex_header.py
Comment (ronaldmaj) [inline scripts/gn/parse_rinex_header.py:1]: This is really awesome and something we haven't had in Python form yet - this will bring a lot of value for people using Ginan with their own RINEX file

**Review by ronaldmaj — COMMENTED** [review]
This is looking really great! I still need to test it out, but from what I've seen it's probably in good enough shape to go in. I might get a few more eyeballs on this as well before then

File: scripts/auto_generate_yaml.py
Comment (aclel) [reply scripts/auto_generate_yaml.py:1]: Mostly no - I found by creating the template yaml all I really needed was the ability to override options in the file itself, rather than generate the file from scratch. I copied the write_nested_dict_value function into my tool. I wanted to come up with a better way - any suggestions? Could import from auto_generate_yaml.py for now?

Comment (ronaldmaj) [reply scripts/auto_generate_yaml.py:1]: Nah, not a worry. This was a pretty old script so hadn't been updated in a while. I was contemplating starting from scratch anyway, so for now you can leave the function you took out in your scripts

File: scripts/gn/README.md (hunk 2 above)
Comment (aclel) [reply scripts/gn/README.md:9]: Awesome, I'll take a look at `auto_download_PPP` and we can look at incorporating it in near future

File: scripts/gn/download_rinex_deps.py
Comment (aclel) [reply scripts/gn/download_rinex_deps.py:1]: Sounds good - next step is to merge and refine

File: scripts/gn/README.md (hunk 2 above)
Comment (ronaldmaj) [reply scripts/gn/README.md:9]: Actually, I realise now that it hasn't entered the GitHub repo yet, so maybe I'll send you a copy of that now and you can have a look at it from that version

File: scripts/gn/README.md (hunk 1 above)
Comment (aclel) [reply scripts/gn/README.md]: Just added some instructions to the README

File: scripts/gn/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+One the pea has finished running, open up the sinex file in the `workspace/test/outputs` folder to see the estimated position in the `SOLUTION/ESTIMATE` block.
+
+**Note**: This command is a work in progress and currently only supports recent files that you would submit to AUSPOS and state government for inclusion in the national adjustment:
+
+* Only handles RINEX v3
+* Only works for RINEX files that were collected >2 weeks ago (just requires FIN products)
+* Only works on RINEX files collected after 2022-11-27 (long filenames started on CDDIS)
+* Only handles a single RINEX file at a time
+* Only handle downloads from CDDIS
+* Recommendation to use RINEX files with >6 hours of data. The generated yaml configuration has ambiguity resolution is turned off, so while pea will still run with a shorter observation session, the results may be well off.
```
Comment (treefern) [inline scripts/gn/README.md:48]:
```suggestion
* Recommendation to use RINEX files with >6 hours of data. The generated yaml configuration has ambiguity resolution turned off, so while pea will still run with a shorter observation session, the results may be much less accurate.
```
Small typo fix. Change of phrasing to avoid using an idiom for `inaccurate` which may confuse some readers.

File: scripts/gn/main.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+        logging.info("Generating yaml config file")
+        overrides = create_overrides(header, station_alias, workspace_name)
+        yaml_path = write_yaml(target_dir, workspace_name=workspace_name, overrides=overrides)
+
+        logging.info("Ready to run the pea...")
+        return yaml_path
+    else:
+        raise NotImplementedError(
+            "Only static ppp is supported through this interface - include --static --ppp"
```
Comment (treefern) [inline scripts/gn/main.py:207]:
```suggestion
            "Only static ppp is supported through this interface - include --static --ppp "
```
Will concatenate to `-pppYou` as is

File: scripts/gn/parse_rinex_header.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        raise NotImplementedError("Only RINEX v3 is currently supported")
+
+    # Load observations to determine which signals have been observed
+    # by the receiver for each gnss system (GPS, Galileo etc)
+    obs = gr.load(filepath)
+    sys_signals = _get_signals_per_system(obs)
+
+    marker_name = header["MARKER NAME"].strip()
+
+    # TODO: Handle the case when not all the fields are given - can't unpack 3 fields when there are only two values
```
Comment (treefern) [inline scripts/gn/parse_rinex_header.py:80]: I appreciate the call out on this one, given issues I've seen before.
Can we determine in a straightforward way whether the fields are missing, and raise an exception on it?

Comment (aclel) [reply scripts/gn/parse_rinex_header.py:80]: It's hard to tell which fields are missing. It's possible that only the TYPE (middle value) could be provided, and when we split we would only get one value back. It would be hard to tell that we're missing "REC #" and "VERS".

We would need some kind of way to check that each element matches a pattern, ie. check if the TYPE is in a list of valid receivers, does the REC # conform to a certain pattern, is the VERS valid. These would be assumptions and I'm not sure if user data would always follow these patterns.

I'm not sure the best way of handling it at the moment.

Comment (treefern) [reply scripts/gn/parse_rinex_header.py:80]: Right. Yeah hard to handle if less than three values **could be valid**, but telling what is what, and validating it is non-trivial.

I guess !=3 elements: raise `NotImplemented` for now? (Are we actually looking at values though, or just the header?)

I can't quite see how the `TYPE` is being _extracted_ as it is on line 82.
~~Line 69 is also confusing me... ohh~~. I'd suggest fully qualifying `gr` as `georinex`. I misread it as `gn` several times and couldn't figure how anything was working given `rinexheader()` isn't defined locally, but there's a class called `RinexHeader` with no explicit constructor.

File: scripts/requirements.txt [hunk 6]
```diff
@@ -20,4 +20,6 @@ pymongo>=3.7.2
 pytest>=7.1.2
 scipy>=1.7.3
 statsmodels>=0.13.2
-gnssanalysis>=0.0.28
```
Comment (treefern) [inline scripts/requirements.txt:23]: Was this intentional? `gnssanalysis` is leveraged in `download_rinex_deps.py`

File: scripts/requirements.txt [hunk 7]
```diff
@@ -20,4 +20,6 @@ pymongo>=3.7.2
...
+#georinex>=1.16.2
+git+https://github.com/breid-phys/georinex.git # fast processing patch - not yet merged into georinex https://github.com/geospace-code/georinex/pull/85
+-e . # Add local gn
```
Comment (treefern) [inline scripts/requirements.txt:25]: Is this here as a convenience primarily?
I gather it's streamlining things for a better user experience (same effect as running `pip install .` after gathering the above requirements?)

File: scripts/requirements.txt (hunk 6 above)
Comment (aclel) [reply scripts/requirements.txt:23]: Whoops, not intentional. I had replaced it with my local version of gnssanalysis because I made some changes and they weren't merged in yet. Meant to leave it unchanged in requirements.

File: scripts/requirements.txt (hunk 7 above)
Comment (aclel) [reply scripts/requirements.txt:25]: Yep just for convenience

Comment (treefern) [reply scripts/requirements.txt:26]: After some other discussions I realised `-e` apparently _symlinks_, not installs. Still convenient, but maybe a bit of unexpected behaviour we should avoid doing in `requirements.txt`?

File: scripts/gn/parse_rinex_header.py (hunk 5 above)
Comment (aclel) [reply scripts/gn/parse_rinex_header.py:80]: Line 82 was what I was using to parse a specific file that I have that only has TYPE (not the other two elements), but it needs to be more general.

The only element that's being used is TYPE (receiver type), which ends up in `receiver_options: STATION: receiver_type:` in the ginan yaml. Does ginan actually use this information for anything? Or does it just care about the *antenna type* to pull out the correct values from the igs20.atx file?

If it's not using it, then perhaps we don't parse it at all for now? If it is, then could we split the field into its elements and check any of the elements are found in a list of valid TYPEs (not sure where to get this list but I'm sure it would exist somewhere).

It would be nice if we could raise an exception if there are not 3 elements, but I've come across a fair few files that don't have this field fully populated. Trying our best to pull out a valid type might be nicer than asking the user to pad out the field with dummy data so that there are 3 elements to parse.

File: scripts/requirements.txt (hunk 7 above)
Comment (aclel) [reply scripts/requirements.txt]: Agree, see commit [998bc72](https://github.com/GeoscienceAustralia/ginan/pull/62/commits/998bc72b9fac517f391f52fd7fe63c2e3bb5cc46) which removes it from `requirements.txt` and adds a new instruction to the README to do a `pip install .`

File: scripts/gn/parse_rinex_header.py (hunk 5 above)
Comment (treefern) [reply scripts/gn/parse_rinex_header.py:80]: Apparently `receiver_type` hardly gets used. And I haven't found a list of valid `TYPE`s anywhere.
So given it's also hard to parse out robustly, I'd agree we probably shouldn't parse it for now.

File: scripts/gn/README.md [hunk 8]
```diff
@@ -12,8 +12,14 @@ cd /path/to/ginan/scripts
 # Create a new virtualenv
 python -m venv env
 
+# Activate the virtualenv
+. env/bin/activate
```
Comment (treefern) [inline scripts/gn/README.md:16]: Valid, though `source` is _reasonably_ universal and probably clearer.

Comment (aclel) [reply scripts/gn/README.md:16]: Sounds good

File: scripts/gn/parse_rinex_header.py (hunk 5 above)
Comment (aclel) [reply scripts/gn/parse_rinex_header.py]: See [4cedd91](https://github.com/GeoscienceAustralia/ginan/pull/62/commits/4cedd911bd985f36f8690452dbcca2a10d8f0941)

**Comment by treefern** [comment]
Great stuff. 😄
I think the only remaining tweak from my perspective is whether some of the more general utility functions (like the Rinex header parser) move into `gnssanalysis`, but that can happen later on.

**Comment by aclel** [comment]
Thanks for the feedback!😃 yeah keen to keep improving it over time. Moving some stuff into gnssanalysis sounds good

File: scripts/gn/templates/auto_template.yaml [hunk 9]
```diff
@@ -0,0 +1,319 @@
+inputs:
+
+    inputs_root: "."
+
+    atx_files:  [ ../../inputData/products/igs20.atx                    ] # required
+    egm_files:  [ ../../inputData/products/tables/EGM2008.gfc           ] # Earth gravity model coefficients file
+    igrf_files: [ ../../inputData/products/igrf13coeffs.txt             ]
```
Comment (ronaldmaj) [inline scripts/gn/templates/auto_template.yaml:7]: Should be `../../inputData/products/tables/igrf13coeffs.txt` (i.e. in the tables directory)

File: scripts/gn/templates/auto_template.yaml [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+
+
+outputs:
+
+    metadata:
+        config_description:    "187369_v3"
+
+    outputs_root:              ./outputs/
+
+    # gpx:
```
Comment (ronaldmaj) [inline scripts/gn/templates/auto_template.yaml:38]: It would be good to have gpx turned on by default - it's an easier way for users to grab results after the run. The TRACE is ok, but it's a big file and it can be hard to find the details you want
I think this will be my last comment on this (happy for it to go in after this).

**Review by ronaldmaj — CHANGES_REQUESTED** [review]
I've completed the run and this works well! :+1:  
Would be nice not to have to rely on the `s3_filehandler` download as it comes along with a whole bunch of other files, but I think that'll be a point for the next iteration on this.

**Review by ronaldmaj — APPROVED** [review]
Looks good now, I think all issues were addressed
